### ISSUE TITLE
[RE] MEDIRL Maximum Entropy Deep Inverse Reinforcement Learning for Human Social Navigation

### ISSUE THREAD

**Issue by nihalgunu** [ReScience/submissions#79]
**Original article:**
Learning How Pedestrians Navigate: A Deep Inverse Reinforcement Learning Approach
https://par.nsf.gov/servlets/purl/10111828

**PDF URL:**  
https://drive.google.com/file/d/1Ud4husXegrcooQWrjyjaH6UPsevW_7Su/view

**Metadata URL:**  
https://docs.google.com/document/d/1jJngE8_9QGPyu3Q6PxKqrhVrpmbVvS5FblQiwc2DgSs/edit?usp=sharing

**Code URL:**  
https://dagshub.com/ML-Purdue/hackathonf23-Stacks

**Scientific domain:** 
Data-Driven Approach to Human Social Navigation, Human-Robot Interaction within Navigation, Deep Inverse Reinforcement Learning
**Programming language:**  
Python
**Suggested editor:**

**Comment by rougier** [comment]
Thanks for your sublmission, we'll assign an editor soon. @koustuvsinha or @gdetor Can you handle this submission?

**Comment by vingupta22** [comment]
Hey @rougier, I'm the co-author of this paper, is there any update for the review?

**Comment by rougier** [comment]
@vingupta22 Sorry for delay. No update so far.
 @koustuvsinha or @gdetor, can you handle this submission?

**Comment by jinensetpal** [comment]
Hi @rougier, may I review this submission?

**Comment by rougier** [comment]
@jinensetpal Sure, thanks for your help.

**Comment by rougier** [comment]
@ReScience/reviewers We need a second reviewer for Machine Learning / Python article, any takers?

**Comment by gdetor** [comment]
@rougier I can be the second reviewer.

**Comment by rougier** [comment]
@gdetor Great, thank !

**Comment by rougier** [comment]
@jinensetpal @gdetor You can start your review and if you can do it by mid-March that would be great.

**Comment by rougier** [comment]
@nihalgunu @vingupta22 Sorry for such very long delay. Hopefully this can be completed by end of March. Still interested in being published?

**Comment by vingupta22** [comment]
@rougier We are definitely still interested! No worries.

**Comment by gdetor** [comment]
Main Text
=======
The main text is overall well written, clear, and comprehensible. However, I have a few comments.

1. In the first two sections, the authors repeat themselves multiple times. For instance, the three last paragraphs of the introduction repeat the message the authors introduced earlier. 
2. In section 3, where the authors introduce the discount factor, there is a typo (on its distance from the mobile robot). 
3. Equations 3, 4, 5, and 6 need some improvement. Please use the subscripts and superscripts properly (per the original article). 
4. On page 6, there are a few typos: "Given the missing information, we had the" -> we had to | "input and output layers given out data"    -> given our data
5. In algorithm 1, the MAXENTIRL and TRAINMAXENTIRL functions do not convey critical information to the reader. Also, the algorithm is TensorFlow-oriented. I suggest re-writing it more generically. 
6. After equation 8, the authors mention that all models' run time is O(N). What is N?
7. Is that the authors took screenshots of the Tensorboard and used them as figures in the text. However, it would be better to have larger and cleaner vectorized figures. 
8. It would be better if the authors gathered together relevant figures (e.g. even numbered figures) using subplots. It is easier for the reader to compare the results of the different ablation cases simultaneously. 


Source Code
=========

1. Missing LICENSE file
2. The source code needs to be organized better (please see here https://www.dailydoseofds.com/how-to-structure-your-code-for-machine-learning-development/  or https://gist.github.com/ericmjl/27e50331f24db3e8f957d1fe7bbbe510)
3. There is no documentation to support the source code. It would be nice if the authors had their source code documented. 
4. The requirements.txt looks too long (like a pip freeze printout). Please provide clean requirements.txt and a description of the platforms where you have tested the source code already (OS, Python version, GCC compiler, etc).  

> [!CAUTION]
> For some reason, I cannot clone the repository, so I cannot run the code.

**Comment by vingupta22** [comment]
Thank you for the feedback, @jinensetpal do you have any additional comments or shall Nihal and I get started on making these changes?

**Comment by jinensetpal** [comment]
@vingupta22 Haven't read the first review so as to get a fresh pass on your paper - I'll make sure to go through it after my review is ready to ensure you don't have overlapping feedback. I will add my review on 3/13 or 3/14.

**Comment by jinensetpal** [comment]
# Overall Review
The work here is a meaningful start. Experiments are well-scoped and training code within individual files is structured well. However, some critical components of the original study remain missing, and a very small portion of the dataset is used for training. Therefore, conclusive statements cannot be made from this replication just yet.

## Major Considerations:
1. >The lack of substantial information about the Social Affinity Map (SAM) feature map added to the complexity of our replication efforts.

The paper references [22] http://vision.stanford.edu/pdf/alahi14.pdf for SAM, which includes a rigorous definition.

2. > Unfortunately, the paper did not provide a reference or access to the dataset used, which further complicated our task.

Section IV. A. of the original paper includes a reference [16] to the publication introducing the dataset along with a functioning link (http://www.irc.atr.jp/crest2010_HRI/ATC_dataset) to download it. Furthermore, in the replication study, the authors also reference dataset statistics from ATC.

3. **Scaling this replication study is necessary:**
  - I computed the loss of the network with no parameter updates (random initialization) and got an error (avg entropy + alignment) of ~1.671. Despite extending the number of epochs to 50, I was only able to get a minor improvement in training error of ~1.342.
  - This likely a result of using just a small subset for training:
> 100 lines for training and 40 lines for testing. We did this to adjust the dataset to be suited for the lack of computational power we had available to us for this study.

Running training runs on such a small fraction (paper mentions the full dataset to be 15 million samples) of the dataset likely results in training performance that is unrepresentative.
4. Equations 3 & 4 need to correct notation with subscripts, as D and θ are both variables defined separately. Understanding that the first and second terms of the derivative corresponded to the first and second terms of the loss function was not inherently clear to me.
5. Regarding performance improvements from the exclusion of the vertical dimension:
>traversing space will reduce the dimensionality of the issue making it easier for the model to understand.
>This is likely because the humans within the environment are not vertically moving and this additional dimension just leads to excess unnecessary error.

If this is the case, the feature representing the vertical state dimension would be constant, and effectively function as an implicit bias. While this increases the number of parameters, it should not affect predictions unless the collection of data is noisy. Additional studies demonstrating high variance in what should be a static feature for this dataset is necessitated.
6. The exclusion of SAM in this work is concerning, as this significantly constraints what can be inferred from data. In the original paper, it is mentioned:
> An MEDIRL algorithm was developed to learn the reward function from expert demonstrations, and feature vectors are selected utilizing SAM representing social affinity in human navigation behaviors

Showing the importance of SAM. Currently, the only features provided to the network in the replication study are positions and velocities, suggesting a large part of the learning may come from linearly interpolating trajectories, rather than modelling social dynamics as part of prediction.

## Minor Considerations:
1. IMO the vocabulary here probably needs to be grounded. Examples:
  - "groundbreaking work"
  - "serves as a testament to the pursuit of knowledge, with the ambition to redefine and fortify the pathways to socially intelligent navigation."
2. Under reproducibility summary, re-experimentation process" point 4 is a re-iteration of point 2.
3. Tense should be in present / past rather than future. Examples:
  - Our objective is to -> Our objective was to 
  - result of our ablation studies will provide -> result of our ablation studies provides
4. Minor Typos:
  - keeping the bigger one of 4028 -> keeping the bigger one of 4096
  - removed the RelU -> removed the ReLU

## Implementation Notes:
1. `git clone` does not work, workaround to this was downloading as a zip file, probably needs to be fixed on the DagsHub side.
2. Make remote mlflow logging optional, by default it will log experiments locally.
3. Probably a good idea to document the python version used through this project.
4. `requirements.txt` is bloated (e.g. torch dependencies are included but never used), using `pipreqs` I obtained one that only states dependencies that explicitly need to be mentioned.
```
dagshub==0.5.9
keras==2.13.1
matplotlib==3.10.1
mlflow==2.20.4
mlflow_skinny==2.20.4
numpy==1.26.4
pandas==2.2.3
scikit_learn==1.6.1
tensorflow==2.13.1
dvc==3.59.1
```
5. make pathfiles relative using pathlib, so that one does not have to change hardcoded filepaths and can run code without additional modification.
6. `model.h5` is not present in the DVC cache and is also not generated as a part of the replication study, if that is needed, where should be clarified, and if not it should be removed.
7. There is a lot of code repetition, and in cases includes training functions (see `reproduced_model_runner.py`) in evaluation code that is never called, those could definitely be trimmed out and a function.
8. Getting setup with the repository also requires knowledge of DVC, I'd probably add a note to run `dvc pull -r origin` to obtain the data somwhere in the README.

**Comment by rougier** [comment]
@jinensetpal Thanks for your very detailed review.
@vingupta22 You can start revising your article and please post here what you've changed.
